FLOE-437 + FLOE-439 + FLOE-441

waharnum opened the pull request:
This PR covers three separate issues:

*Sonifier implementation*: https://issues.fluidproject.org/browse/FLOE-437

*Better test coverage for existing features*: https://issues.fluidproject.org/browse/FLOE-439

*Connecting the pie's description and title SVG elements to the inline edit components*: https://issues.fluidproject.org/browse/FLOE-441

New files in the `src/lib/flocking` directory don't need to be reviewed - they're dependencies for the data sonification from the http://flockingjs.org/ library.

cindyli commented on demos/index.html:
```diff
@@ -62,7 +72,28 @@
                                 width: 400,
                                 height: 400
                             }
-                        }
```
The options block for instantiating "floe.chartAuthoring.demo" is getting longer and scary. Might be nice to rearrange some of those. I'm thinking options like dataEntryPanel.strings, pieChart.pieChartOptions.xxxDisplayTemplate can be moved into components where they are defined. Other options like for templateLoader, pieChart.pieChartOptions.colors and initialData can be moved to chartAuthoring-demo.js. Let me know if it makes sense.

waharnum replied:
@cindyli: By "moving into components", do you mean maybe some of these options should become the default options of the actual components involved in the chartAuthoring component, rather than being configured this way for the demo?

cindyli replied:
Yes, exactly.

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -66,6 +68,12 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
                                     text: "floe-inlineEdit-text"
+                                },
+                                listeners: {
+                                    "afterFinishEdit": {
```
Add namespace for listeners.

cindyli replied:
The listener can be simplified to:

```
"afterFinishEdit.modelChange": {
    func: "{pieChart}.applier.change",
    args: ["pieTitle", "{arguments}.0"]
}
```

Same comment applies to the listener at line 95

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -82,6 +90,12 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
                                     text: "floe-inlineEdit-text"
+                                },
+                                listeners: {
+                                    "afterFinishEdit": {
```
Same as above, add namespace.

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -102,7 +116,8 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
                                         func: "floe.chartAuthoring.dataEntriesToPieChartData"
-                                    }
+                                    },
+                                    backward: "never"
```
This line is un-necessary since "fluid.transforms.free" doesn't support the backwards transformation.

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -66,6 +68,12 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
                                 },
                                 styles: {
```
Line 66-71 are common options used by both fluid.inline components. To reduce the repetition, we can defined a top option block such as "inlineEditCommonOptions" that contains these lines then use distributeOptions to distribute them into all "fluid.inlineEdit".

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -121,6 +136,31 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+                                    "dataSet": "{dataEntryPanel}.model.dataSet",
+                                    // We relay currentlyPlayingData and isPlaying to the overall component so that sonification play events can be used to change the rest of the interface
+                                    currentlyPlayingData: "{chartAuthoring}.model.currentlyPlayingData",
+                                    isPlaying:
```
line 155, 156 should be all in one line? remove the empty line 157.

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -152,11 +196,24 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+            currentlyPlayingData: null
+        },
+        modelListeners: {
+            currentlyPlayingData: {
```
Can be simplified to `currentlyPlayingData: "{that}.highlightPlayingData"`.

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -152,11 +196,24 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
             defaultTitleText: "Enter Chart Title",
-            defaultDescriptionText: "Enter Chart Description"
+            defaultDescriptionText: "Enter Chart Description",
+            dataPlayingHighlightClass: "floe-ca-currently-playing"
```
Should this line more belong to "styles" option instead of in "strings"?

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -188,6 +245,72 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+            // Highlight pie chart slices / legend rows as they play
+            if(currentlyPlayingData !== null) {
+
+                var activeRow = rows.filter(
```
The action taken for rows and slices are pretty much same, can we create a function to share?

waharnum replied:
I created some filtering functions in d3Utils.js to enable filtering D3 element sets by a particular data ID, then refactored the highlightPlayingData function to use them.

We should be able to reuse them in other scenarios where we want to be able to highlight or otherwise operate on sets of data-bound elements based on their data IDs.

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -188,6 +245,72 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+    };
+
+    // Adds and removes highlight
+    floe.chartAuthoring.highlightPlayingData = function(that) {
```
I think the highlighting should happen in pieChart component instead of at the top level. What do you think?

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -237,6 +360,24 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
             $(this).find(dataEntryValueSelector).val(currentData.value).trigger("change");
         });
+
+        that.events.onUpdateDataEntryPanel.fire();
```
Might be more accurate to name the event as `onDataEntryPanelUpdated`.

waharnum replied:
@cindyli - I would be concerned that would imply the event is fired whenever the data entry panel is updated, such as by user action; in this case it's actually fired when the floe.chartAuthoring.updateDataEntryPanel function that updates the data entry panel and the corresponding pie chart from a dataset completes.

That said, I think that function itself is probably badly named - propose renaming it to "updateInterfaceFromDataSet", and will rename the event accordingly. Does that sound workable?

cindyli replied:
Right, your concern is reasonable. In that case, shall we use a event name that describes what had happened when this event is fired, such as "onDataApplied" or "onDataAppliedToEntryPanel".

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -237,6 +360,24 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+        that.events.onUpdateDataEntryPanel.fire();
+    };
+
+    floe.chartAuthoring.addSonificationControlListeners = function(that) {
```
My first instinct at seeing "add...Listeners" was to add event listeners. What do you think of names like "addSonificationButtonHandler" or "addSonificationControllerHandler"?

waharnum replied:
@cindyli - works for me. I should remember that "listener" has a very specific meaning in Infusion and avoid using it in the more general sense in favour of "handler".

cindyli commented:
Can you add year 2016 to the copyright line on files that have been modified? Thanks.

cindyli commented on src/js/electricPianoBand.js:
```diff
@@ -0,0 +1,120 @@
+
+    "use strict";
+
+    fluid.defaults("floe.chartAuthoring.electricPianoBand", {
```
@colinbdclark might be the better person to review this file.

colinbdclark replied:
If it's possible, we should have someone else cast an eye on this file, too, since I wrote it. :)

cindyli replied:
I looked it over, no comments.

waharnum replied:
What would probably be helpful at some point (not urgent) would be to annotate this file so it can be used as more of a model synth definition for those largely unfamiliar with how Flocking works or with the particularities of audio synthesis to create their own synths. Maybe @colinbdclark & I can do this together at some point.

I could eventually see something along the lines of an approach to generating the synth definitions Flocking needs from a simple user interface that has pre-defined instrument definitions, etc etc. I'm thinking waaaay back to when I used to play around with MIDI files in the early 90s - just changing the instruments used in MIDI files I found online to hear what things sounded like on a tuba vs. a piano, etc etc.

cindyli commented on src/js/legend.js:
```diff
@@ -273,6 +273,7 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
         that.events.onLegendCreated.fire();
     };
 
+    // TODO: refactor into a utility function, now used by sonifier
```
Can we fix it?

waharnum replied:
@cindyli Does it make sense to make a separate `chartAuthoringUtils.js` file? We've got `d3Utils.js`, but an ascending sort isn't D3-specific. 

Alternately, maybe `d3Utils.js` should become `chartAuthoringUtils.js`.

cindyli replied:
I think it makes sense to create a separate `utils.js` file. As it will be under `chartAuthoring` folder, using `utils` should be ok. 

I'd prefer still keep d3 related operations in a separate file so that we know the dependency. If someday we decide to use another drawing library, it saves us from the cherry picking in the code.

waharnum replied:
Good points about dependency isolation - I'll create a separate `utils.js`

amb26 commented on src/js/chartAuthoring.js:
```diff
@@ -237,6 +347,24 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+        var playButton = that.locate("sonifierPlay"),
+            stopButton = that.locate("sonifierStop");
+
+        playButton.click(function(e) {
```
These listeners form a pretty stereotypical pattern. I suggest drawing up some configuration accepted by the component and encoding this declaratively.

amb26 commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,402 @@
+        return durations;
+    };
+
+    floe.chartAuthoring.sonifier.getSonificationEnvelopeDurations = function(units, unitDivisor, envelopeDurationConfig) {
```
These functions all follow a common pattern, please refactor them

amb26 commented on src/lib/flocking/ugens/envelopes.js:
```diff
@@ -0,0 +1,1101 @@
+/*
```
We should source this material from npm or bower rather than committing it wholesale

colinbdclark replied:
I will be removing Bower support from Flocking 0.3.0, so I recommend sourcing it from npm. Since there has been a long time between Flocking 0.1.0 and the as-yet-unreleased 0.2.0 version, I recommend using the latest commit hash. So the package.json dependency would look something like this:

```
"flocking": "colinbdclark/flocking#fba0738886b3b6aa1f6292863d2fa50a39cd47e3"
```

In the future, I will investigate using Fluid's npm publishing tools to more frequently publish development snapshots.

amb26 replied:
Cool - in the meantime, can we get both flocking and infusion out into npm references to get them out of this diff? I assume that we can generate the infusion build using an npm "postinstall" step.

amb26 commented on src/js/pie.js:
```diff
@@ -21,19 +21,25 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
 
     fluid.defaults("floe.chartAuthoring.pieChart.pie", {
-        gradeNames: ["floe.chartAuthoring.totalRelaying", "floe.d3ViewComponent",  "autoInit"],
+        gradeNames: ["floe.chartAuthoring.valueBinding","floe.chartAuthoring.totalRelaying", "floe.d3ViewComponent"],
```
Something about the linting configuration is incorrect, as you can see if you try to introduce deliberate errors into this file - despite it reporting "190 files lint free", it seems that somehow, the project implementation files themselves are escaping linting? Please could you look into this..... cheers

waharnum replied:
@amb26 - what errors are you introducing? I get the following when inserting an indentation error and a loose equality vs. strict:

```
Linting src/js/pie.js ...ERROR
[L24:C5] W015: Expected 'gradeNames' to have an indentation at 9 instead at 5.
    gradeNames: ["floe.chartAuthoring.valueBinding","floe.chartAuthoring.totalRelaying", "floe.d3ViewComponent"],
[L214:C37] W116: Expected '===' and instead saw '=='.
                return typeof (d) == "object" ? d.value : d;
```

amb26 replied:
Hi @waharnum - the very form of those linting messages looks puzzling and unexpected (e.g. including text like "W015" and "[L24:C5]") - and not of the kind generated by the linting task - the ones I am referring to are the ones output from the grunt plugin by running the task "grunt lint". I introduced both indentation and blatant syntax errors but got no output from the standard task - what are you running in order to generate the transcript you got above?

waharnum replied:
@amb26, I'm also running `grunt lint` - but it's possible I'm having a versioning issue with the node_modules that enable the linter or otherwise (it wouldn't be the first time). 

I ran `npm update` to make sure I had latest versions of everything in `package.json` and see this output now when running `grunt lint` after introducing the same indentation / loose vs. strict error:

```
Running "jshint:all" (jshint) task

   src/js/pie.js
     24 |    gradeNames: ["floe.chartAuthoring.valueBinding","floe.chartAuthoring.totalRelaying", "floe.d3ViewComponent"],
             ^ Expected 'gradeNames' to have an indentation at 9 instead at 5.
    214 |                return typeof (d) == "object" ? d.value : d;
                                             ^ Expected '===' and instead saw '=='.

>> 2 errors in 33 files
Warning: Task "jshint:all" failed. Use --force to continue.

Aborted due to warnings.
```

amb26 replied:
It's possible but it seems to be more than a versioning issue since for me these files are not being linted at all. I just did a git clone and an npm install - you can simulate the same with an rm -rf node_modules and npm install - see what it turns up!

amb26 replied:
Oh, really sorry for these (mostly) false alarms. Since I am on my laptop today everything is just a bit too far away to see. Embarrassingly I've just been running linting on the wrong project. But the main thing which brought this up visually was the fact that somehow the option "white": true appears to have gone missing from the core .jshintrc file in infusion. However, with this in, the core framework itself blatantly fails to lint. I'll need to follow up with @jobara and others what the exact history was here, since I'm sure that one of the main reasons for persisting with our incredibly ancient pinned jshint version was precisely to get whitespace linting. Interestingly it is here in the GPII main repo: https://github.com/GPII/universal/blob/master/.jshintrc#L23

waharnum replied:
@amb26 - no worries. it sounds like this at least serendipitously surfaced an issue with the linting in the core framework. I was slightly confused because if there's one thing I'm sure of, it's that my linter for this project complains at me when I don't indent properly due to https://gist.github.com/waharnum/92ffbe9de590e164de30 :)

amb26 replied:
Cheers, that gist is an awesome guide! Sorry for the waste of time. The formatting of your linting messages is still puzzling though, mine appear as follows:

   Gruntfile.js
     15 |module.exports = function(grunt) {
                                  ^ Missing space after 'function'.

waharnum replied:
Mine look the same after an `npm update` - I suspect it was an older version of one of the modules involved with a different style of linting messages.

cindyli commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,402 @@
+                        }
+                    },
+                    listeners: {
+                        "{that}.events.onStop": {
```
This line can be simplified to `"onStop.playDataAndQueueNext":`

amb26 replied:
Nonetheless, it should have a namespace!

cindyli commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,402 @@
+    // convert it to an array of objects in the style used by the sonification components,
+    // maintaining object constancy by using the dataEntry object name as the key
+    floe.chartAuthoring.sonifier.dataEntriesToSonificationData = function(that) {
+        var unitDivisor = 10;
```
Does it make sense to turn this variable into a component option?

colinbdclark commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,402 @@
+                // Always stop the flocking environment
+                flockingEnvironment.stop();
+                // Reset the bus manager
+                flockingEnvironment.busManager.reset();
```
You should be able to call <code>reset()</code> on the Flocking environment directly, which will stop it and reset all of its underlying components and state, including the Bus Manager. This API is influx, so try to avoid accessing components of the Environment wherever possible.

Though I wonder if this is somewhat heavy-handed? I imagine that the sonifier should just stop the Flocking environment and start it again later if needed. Are you resetting the Bus Manager in order to flush samples out of the busses or because of something audible you've encountered?

waharnum replied:
@colinbdclark - can try and document these issues more thoroughly later, but off the top of my head I found the following when working to implement the stop functionality. It's possible / likely that the issues I encountered were because of problems with the code's approach to using Flocking that you highlighted in comments below:

**Attempt 1 - flockingEnvironment.stop()**

This would cause successive attempts to play the sonification following a stop to fail with a "cannot acquire bus" error.

**Attempt 2 - flockingEnvironment.reset()**

This would cause successive attempts to play the sonification to fail without an obvious reason (no error to console).

**Current implementation - flockingEnvironment.stop() + flockingEnvironment.busManager.reset()**

This was the one approach I found that would stop the current play while still allowing successive plays.

Once I refactor the Flocking usage as you recommend below, I'll refactor the stop functionality as well and see if the less heavy-handed approach works.

colinbdclark replied:
I'll be curious to see if, after your refactoring, you still encounter an issue with stopping or resetting the Environment. If so, it's definitely a bug in Flocking and I will fix it.

If you do still see problems and have a spare moment, don't hesitate to [file a Flocking ticket](https://github.com/colinbdclark/Flocking/issues) about it.

colinbdclark commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,402 @@
+    floe.chartAuthoring.sonifier.stopSonification = function(that) {
+
+        if(that.model.isPlaying) {
+            try {
```
Why is this wrapped in a <code>try/catch/finally</code>? Which of these functions will potentially throw, and under what circumstances?

colinbdclark commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,402 @@
+    // execute a sonification
+
+    floe.chartAuthoring.sonifier.startSonification = function(that) {
+        if(!that.model.isPlaying) {
```
Flip this to avoid excess indentation:

```
if (that.model.isPlaying) {
    return;
}
```

colinbdclark commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,402 @@
+
+            flockingEnvironment.start();
+
+            fluid.defaults("floe.chartAuthoring.dataPianoBand", {
```
Why are we defining this component each time the sonification is started? Why can't it be defined once, ahead of time?

waharnum replied:
I think this is an inadequately refactored legacy of the "get something that works quickly" code from last November. I'll see about reimplementing this so that we only define it once.

amb26 replied:
As a general strategy note, fluid.defaults blocks should never be issued from functions - unless they generate fresh, uniquely named "nonce grades"

colinbdclark commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,402 @@
+                }
+            });
+
+            var dataPianoBand = floe.chartAuthoring.dataPianoBand();
```
Why are we manually instantiating this component here? Can it be refactored as a <code>createOnEvent</code> component?

waharnum replied:
See comment above - I'll refactor to a `createOnEvent` component as part of the component tree.

colinbdclark commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,402 @@
+            var dataPianoBand = floe.chartAuthoring.dataPianoBand();
+
+            // Add the synth to the model
+            that.applier.change("synth", dataPianoBand);
```
Components aren't suitable as model material. Can this component be defined as a subcomponent of the Sonifier?

cindyli commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,402 @@
+
+    floe.chartAuthoring.sonifier.getSonificationEnvelopeValues = function(envelopeDurations, envelopeDurationConfig, envelopeValuesConfig) {
+        var envelopeValues = fluid.transform(envelopeDurations, function(duration) {
+            if(duration === envelopeDurationConfig.divisorDuration || duration === envelopeDurationConfig.remainderDuration) {
```
line 238-242 can be simplified:

```
var isDurationMatched = duration === envelopeDurationConfig.divisorDuration || duration === envelopeDurationConfig.remainderDuration;
return envelopeValuesConfig[isDurationMatched ? "openValue" : "closedValue"];
```

waharnum replied:
@cindyli, that approach of using a ternary to figure out which key to use is awesome. :+1:

cindyli commented on tests/html/totalRelaying-Tests.html:
```diff
@@ -0,0 +1,36 @@
+    </head>
+
+    <body id="body">
+        <h1 id="qunit-header">Chart Authoring Tests</h1>
```
The header needs to be changed.

cindyli commented on tests/html/totalRelaying-Tests.html:
```diff
@@ -0,0 +1,36 @@
+
+        <!-- Test HTML -->
+
+        <div class="floec-chartAuthoring"></div>
```
This selector doesn't seem to be used in the test js.

cindyli commented on tests/js/chartAuthoringTests.js:
```diff
@@ -155,8 +160,50 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+        dataEntries.each(function(idx) {
+            var expectedData = floe.tests.chartAuthoring.updateDataSet[idx];
+
+            if(expectedData !== undefined) {
```
If all data entries are expected to be updated, would it be a case that expectedData is `undefined`?

waharnum replied:
@cindyli It can be undefined because the length of expectedData may be less than the length of dataEntries (which is always 5). That said, I don't think this is quite the right approach to the test and would miss certain cases - I'll refactor.

cindyli commented on tests/js/chartAuthoringTests.js:
```diff
@@ -155,8 +160,50 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+            var expectedData = floe.tests.chartAuthoring.updateDataSet[idx];
+
+            if(expectedData !== undefined) {
+                jqUnit.assertEquals("Testing updated dataEntryPanel UI - displayed label at position " + idx + " matches updated data set label", $(this).find(dataEntryLabelSelector).val(), String(expectedData.label));
```
Can we use `that.locate()` instead of `$(this).find()` at this line and the line below?

cindyli commented on src/js/d3Utils.js:
```diff
@@ -23,6 +23,42 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+        try {
+            var id = d.id !== undefined ? d.id : d.data.id;
+            return id !== undefined ? id : undefined;
+        } catch (e) {
```
Isn't this `catch` block has been caught at line 34?

amb26 replied:
This catch block is unnecessary

waharnum replied:
It was an awkward construction, but actually necessary to avoid a possible error when trying to access d.data.id (d.data might not exist). I restructured into a series of ternary operators so try/catch isn't necessary.

cindyli commented on tests/js/pieTests.js:
```diff
@@ -37,6 +37,23 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+
+    // Test that when model.pieTitle is changed, the corresponding DOM
+    // element is also updated
+    floe.tests.chartAuthoring.testTitleBinding = function (that) {
```
`floe.tests.chartAuthoring.testTitleBinding()` and `floe.tests.chartAuthoring.testDescriptionBinding()` are quite similar, can you refactor them into one.

cindyli commented on tests/js/sonifierTests.js:
```diff
@@ -0,0 +1,135 @@
+
+        that.playSonification();
+
+        jqUnit.assertTrue("sonificationQueue has items in it", that.model.sonificationQueue.length === 2);
```
Shall contents of queued items be verified too?

waharnum replied:
Good idea - added a test for this.

amb26 commented on src/js/d3Utils.js:
```diff
@@ -23,6 +23,42 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+
+    // Given a selection of D3 elements and an ID, returns only the elements
+    // matching that ID
+    floe.d3.filterById = function(d3Selection, currentlyPlayingDataId) {
```
These two functions are identical

amb26 replied:
Not QUITE identical, I see - now I've looked at the above implementation in more detail, I see that these utilities are involved as part of the general strategy of "using our view as a model" which should be factored away. We should track the relationships between playing data, rows and slice ids at the model level of the application rather than storing it in the DOM.

waharnum replied:
@amb26 I've done a first pass at refactoring (still working on some of the pieces) to a model-level approach for everything except the D3-specific functionality that's needed to filter and select DOM elements based on D3's own model of binding data to DOM elements. I think this fits with what we all talked about last year around how and when we make use of D3 in these components.

amb26 commented on src/js/chartAuthoring.js:
```diff
@@ -185,9 +232,59 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+            // Highlight pie chart slices / legend rows as they play
+            if(currentlyPlayingData !== null) {
+
+                var currentlyPlayingDataId = currentlyPlayingData.id;
```
All of this code should be modelised rather than banging on the view layer directly

waharnum replied:
@amb26 i'm not 100% clear on what that would mean in this context (where we're making use of D3's selection/filtering etc), can you elaborate?

amb26 replied:
Hi @waharnum - so what I mean is a model of the form

```
activeRows: {
    <id>: true/false
},
activeSlices: {
    <id>: true/false
}
```

where id holds some form of unique id representing the row slices - and the logic for updating these based on currentlyPlayingData.id is managed purely at the model level.
The code to add/remove classes from the corresponding rows and slices is then positioned as model listeners to activeRows/activeModels

waharnum replied:
@amb26 - that makes perfect sense, thanks.

amb26 commented on src/js/electricPianoBand.js:
```diff
@@ -0,0 +1,120 @@
+        }
+    });
+
+    floe.chartAuthoring.carrierSynth.additiveExpander = function (phases, ugenDef) {
```
This function is interesting and could do with some comments. In what sense is it "additive" since it appears rather to "explode" an array of phase values into a supplied ugen?

colinbdclark replied:
It is poorly named. It expands a single unit generator definition into multiple unit generator definitions, hardcoded to parameterize the unit generator's phase. It is "additive" in the DSP sense, producing unit generators from a "template" for additive synthesis.

amb26 replied:
What is its effect, auditorily speaking? And what is there about its output that couples it to use for additive synthesis? Is it just an array that "someone in the know" will know needs to be used for ugen.sum later?

colinbdclark replied:
The audible effect is that you have three sine tones running out of phase with each other, creating a more complex "reedy" tone. It's a very, very rudimentary model of an electric piano like a Fender Rhodes.

Yes, this expander's output is just an array of unit generators that someone will have to know what to do with.

I was thinking that perhaps we should turn this into an expander that:
1. Can take an array or hash of arrays representing values that should be used to produce unit generators for additive synthesis
2. Automatically wrap these in the necessary <code>flock.ugen.sum</code> unit generator definition

Does that seem sensible and address your concern?

amb26 replied:
Sounds very good to me!

amb26 commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,402 @@
+        var sonificationData = floe.chartAuthoring.sonifier.unitDivisorSonificationStrategy(that, unitDivisor);
+        sonificationData.sort(floe.chartAuthoring.pieChart.legend.sortAscending);
+        that.applier.change("sonifiedData",sonificationData);
+        that.events.onDataEntriesConvertedToSonificationData.fire();
```
Is there anyone who would listen to this event rather than simply listening to changes on "sonifiedData"?

waharnum replied:
Probably not - the code itself doesn't do so in any place. I'll remove it.

amb26 commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,402 @@
+        var numberRemainders = value % unitDivisor;
+        var divisorArray =[];
+        var remainderArray = [];
+        for(var i=0; i <numberDivisors; i++) {
```
Whitespace is all over the place here - could I suggest we actually do revive the "white: true" option that should probably be in the .jshintrc file (and in the core framework too)

waharnum replied:
@amb26 I'll add it back in and fix up the whitespace on this project's files.

waharnum replied:
@amb26 thanks to / in spite of the power of regular expressions, I've now corrected 419 whitespace errors after turning on `white: true` in .jshintrc

amb26 commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,402 @@
+
+    floe.chartAuthoring.sonifier.getSonificationNoteValues = function(units, unitDivisor, noteValueConfig) {
+        var durations = fluid.transform(units, function (unit) {
+            if(unit === unitDivisor) {
```
return noteValueConfig[unit === unitDivisor ? "divisorValue" : "remainderValue"]

amb26 commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,402 @@
+        return divisorArray.concat(remainderArray);
+    };
+
+    floe.chartAuthoring.sonifier.getSonificationNoteDurations = function(units, unitDivisor, noteDurationConfig) {
```
These two functions are identical

cindyli commented on tests/js/totalRelayingTests.js:
```diff
@@ -0,0 +1,81 @@
+    // Should always be 100 (100%)
+    floe.tests.chartAuthoring.totalRelaying.expectedPercentageValue = 100;
+
+    jqUnit.test("Test total relaying for pie/legend-style dataset", function () {
```
These 2 tests are same with the exception of the input data set. Can you create a function to share? Thanks.

amb26 commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,402 @@
+
+        // We shouldn't use the gap if this is the first call for a dataset
+        var pause = noGap? delay : delay+gapDuration;
+        if(fluid.textToSpeech.isSupported()) {
```
Factor this out into a listener in a "speech-aware" grade which listens to a suitable event rather than embedding it in the core implementation in an "if" block

amb26 commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,402 @@
+
+    // Recursion function called from floe.chartAuthoring.sonifier.processSonificationQueue
+
+    floe.chartAuthoring.sonifier.playDataAndQueueNext = function(that) {
```
Interesting - @colinbdclark  - do we imagine that future versions of flocking might have support for such implementations in an automatic and declarative way - perhaps with its "scheduler"? What do we imagine they might look like?

waharnum replied:
@amb26 just speculative, but I imagine a fully declarative scheduler might look similar to `fluid.test.testCaseHolder` - they're both handling some similar challenges of ensuring a series of mixed synchronous and asynchronous events happen in a certain sequence. That might be wildly off base, but it's a thought that occurs to me.

I'd characterize what I've done here as a first pass at a well-behaved scheduler - still too imperative, but hopefully going in the right direction. I'd eventually (maybe) see something like a scheduler grade that works in a similar manner to the queue processing approach I've used here, where a collection of objects with attached behaviours is fed one by one into a processing queue through a combination of events and listeners - hence why I say this might look similar to `fluid.test.testCaseHolder`.

colinbdclark replied:
There is a some rudimentary capacity to declaratively schedule things using Flocking's scheduler, which was implemented hastily in a motel room in Boulder a few years ago. However, it's wildly insufficient, especially for this kind of "mixed scheduling" where different types of actions (scheduling changes on a synth, playing back speech, etc.) are being interleaved. Rationalizing this kind of thing in a fully declarative form is a job for BERGSON 2.0!

"There is no state of mind, however simple, which does not change every moment, since there is no consciousness without memory, and no continuation of a state without the addition, to the present feeling, of the memory of past moments. It is this which constitutes duration."

amb26 replied:
"How can I tell that the past isn't a fiction designed to account for the discrepancy between my immediate physical sensations and my state of mind?"

cindyli commented:
The play and pause buttons are keyboard navigable but not controllable, which means pressing space bar or enter would not start or stop the play.

waharnum commented:
@cindyli pressing enter in Chrome for me allows me to interact with the start/stop controls - space doesn't work (enter works because they're <a> tags, so support that default behaviour). Should we support space as well?

cindyli commented:
Thanks for pointing out ENTER works. Proved after trying again.

I think we eventually would like to support the space since it's a common control for video or audio players. We can create another JIRA for it since this pull request is already big enough.

waharnum commented:
@cindyli created that issue as https://issues.fluidproject.org/browse/FLOE-449

amb26 commented on src/js/chartAuthoring.js:
```diff
@@ -123,6 +125,26 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+                                listeners: {
+                                    "onCreate.escalate": {
+                                        funcName: "{chartAuthoring}.events.onSonifierReady.fire",
+                                        priority: "last"
```
What listener does this listener need to beat? Can it be given a constraint-based priority to make this explicit?

waharnum replied:
I think this was a legacy from when the sonifier had a different createOnEvent. It's not an issue now that it waits until after the `onPieChartReady` event fires. Will remove priority.

amb26 commented on src/js/chartAuthoring.js:
```diff
@@ -237,6 +334,24 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+        var playButton = that.locate("sonifierPlay"),
+            stopButton = that.locate("sonifierStop");
+
+        playButton.click(function(e) {
```
Consult @jobara on what we currently have for a declarative event binding system. If we can't locate something suitable, we can import the following sketch:

https://gist.github.com/amb26/e25255f43d36d08a8943

waharnum replied:
@amb26 I took the same approach that @cindyli recommended previously for the `bindResetClick` listener. Binding the events now looks like this and doesn't require a separate function:

```
            "onToolReady.bindPlayClick": {
                "this": "{that}.dom.sonifierPlay",
                "method": "click",
                "args": ["{chartAuthoringInterface}.sonifier.playSonification"]
            },
            "onToolReady.bindStopClick": {
                "this": "{that}.dom.sonifierStop",
                "method": "click",
                "args": ["{chartAuthoringInterface}.sonifier.stopSonification"]
            }
```

If we'd like a full declarative event binding system, maybe that should be a separate JIRA?  In the current case they aren't particularly complicated handlers & can be implemented declaratively as per above.

amb26 replied:
Cheers, @waharnum, that seems reasonable

waharnum commented:
I've completed a first pass at refactoring the synth to be part of the sonifier component tree, as suggested by @colinbdclark above. This seems to work fine, and the sonification tests and behaviour pass. I also updated the `stopSonification` function per @colinbdclark's suggestions - also seems to work well.

The thing is, this seems to have introduced an error into the chartAuthoring tests and added an extra test to the `fluid.test.testCaseHolder`. As far as I can tell these are both related to the `clearComponent` function of `fluid.instantiator`, so maybe this is related to Infusion failing to clean up the Flocking synth at the end of the test sequence somehow? I'm a little mystified - no error gets thrown on the demo page, so this seems to be occurring specifically in the test context, and only after the synth is made part of the component tree of the sonifier.

This is the commit hash where this issue begins occuring (first commit of the refactor):
- https://github.com/waharnum/chartAuthoring/commit/05d4326c1858cd5669c4d300b28d5063519d02fb - you can see that no files have changed except the sonifier and its test 

Here's the chartAuthoring tests from that commit and the one before it via rawgit for easier comparison:
- https://rawgit.com/waharnum/chartAuthoring/8f4d1e1676d6640f27157596e8339c72dec54f1c/tests/html/chartAuthoring-Tests.html (36 assertions, all passing)
- https://rawgit.com/waharnum/chartAuthoring/05d4326c1858cd5669c4d300b28d5063519d02fb/tests/html/chartAuthoring-Tests.html (37 assertions, 1 failure + the failure from the expect() mismatch)

@amb26 / @colinbdclark / @cindyli - any thoughts? I don't want failing tests, but I also can't see an obvious connection between the refactoring and this after about two hours of debugging. I've confirmed that this seems to occur simply through the combination of `fluid.test.testCaseHolder` and the chartAuthoring component including the refactored sonifier - I removed all tests except the first in the sequence and saw it occur.

amb26 commented:
@waharnum - that's correct, by my understanding, flocking depends on various static components (such as the "environment" and "scheduler") which cannot be relied upon to be automatically cleaned up at the end of an IoC testing framework fixture. The natural behaviour at the end of one of these fixtures is just to destroy the fixture's component tree and it assumes that this achieves all cleanup that is required.

Poking around in your aweszome "rawgit" hosting I see that the error is occuring when destroying the component at the following path:

chartAuthoringTest-cclngafb-360.chartAuthoring.chartAuthoringInterface.sonifier.synth.scheduler

the issue appears to be that there is a "timeConverter" which is registered at two paths - 

chartAuthoringTest-cclngafb-360.chartAuthoring.chartAuthoringInterface.sonifier.synth.scheduler.repeatScheduler.timeConverter

and

chartAuthoringTest-cclngafb-360.chartAuthoring.chartAuthoringInterface.sonifier.synth.scheduler.onceScheduler.timeConverter

Unfortunately by this point the "repeatScheduler" has already been destroyed.

I think that this actually represents an Infusion bug as well - since you shouldn't be receiving a "bare diagnostic" like this. Please keep your rawgit urls available since I will try to patch this in the framework, but the only result will be that you would receive a more intelligible diagnostic once I fix the problem, since the "enviro" will still leak.

In theory, calling "enviro.destroy()" in your test fixture's "onDestroy" would be sufficient, but because of the Infusion bug I suspect this might end up being too late - that is, you will probably receive the "Cannot read property of undefined" failure before your onDestroy gets notified. I suggest what you do in this case is to put this at the implementation level in your test, that is as the final sequence element after your "verifyUpdate" listener

waharnum commented:
Thanks @amb26 - I'll try that approach and report back.

waharnum commented:
@amb26, update on the issue:
- enviro.destroy() at various points in the test sequence didn't seem to work. It sometimes moved around the point where the error occured, but it never went away - it either continued to appear, or putting in enviro.destroy() caused the test sequences to fail in other ways

For now, I've worked around the issue by:
- creating an `onSynthNeeded` event in the sonifier
- firing this the first time the synth is actually needed
- configuring the synth with `createOnEvent: "onSynthNeeded"` in the component tree

I think the above is good practice anyway (@colinbdclark recommended it) and it means we get a working component-level synth that doesn't interfere with the chartAuthoring tests - the synth isn't actually created until it's needed to play sound, which the chartAuthoring tests don't do.

It does mean that testing the synth and sonification behaviour in context of the entire chartAuthoring component may be harder, but we aren't currently doing that.

amb26 commented:
Thanks for investigating, @waharnum - this suggests that the Infusion bug is 100%. I'll file it later today and ping this thread. I'm glad you've been able to find a workaround for now.

colinbdclark commented:
It's definitely also a Flocking bug that [an environment can't be created and destroyed entirely within an IoC tree](https://github.com/colinbdclark/Flocking/issues/152). This appears to be leftover from Flocking's pre-IoC adolescence.

I whipped up [this example](https://gist.github.com/colinbdclark/7a6115f840274b722ea4), and it seems to work now that I have fixed the primary bugs in Flocking's master branch.

Alan, if you construct your own Flocking environment instance within your component tree instead of calling <code>flock.init()</code>, does this fix your issue? You'll need to use <code>colinbdclark/Flocking#f67b1a3d7c66cf476192d44bbee6b448adbe57bc</code> for this to work.

amb26 commented:
Hi @waharnum - this bug (the failure during clearComponent) is actually https://issues.fluidproject.org/browse/FLUID-5812 and was fixed in trunk in November. Please update your build of Infusion to get the fix!

waharnum commented:
@colinbdclark did some initial refactoring of the sonification start as discussed - a basic issue is making sure that user interactions with the play control can't cause repeated plays. Can you suggest a better approach than what I've got (`addToSonificationQueue` is basically a gating function that only calls the changeApplier to add to the queue if a play isn't currently happening, and it seems crude)?

colinbdclark commented:
I think that's probably the extent of what you can do, at least until we have some form of declaratively conditional model listeners in Infusion. This refactoring looks much better.

waharnum commented:
I think this is now ready for further review (and is much better due to all the feedback and refactoring). 

One specific thing I've done as part of the overall refactoring has been to split out the sonification strategy code into a separate JS file (`unitDivisorSonificationStrategy`). At the moment this is purely a mixin grade to keep the functions and configuration options in a separate file, but with further refactoring this could be turned into a subcomponent. At this time I'm not 100% sure what the best component structure is to support the concept of "a sonifier can have multiple strategies available to convert datasets into data that can be used by a synth", but having the strategy separated at the file level is a start.

Thanks @colinbdclark / @cindyli / @amb26!

cindyli commented on demos/src/css/chartAuthoring-demo.css:
```diff
@@ -41,9 +41,29 @@ body {
+}
+
+.floe-sonifierPlay img, .floe-sonifierStop img {
+    max-width: 32px;
```
Probably should use "em" for the future UIO integration.

cindyli commented on modified.txt:
```diff
@@ -0,0 +1,6 @@
+./tests/js/chartAuthoringTests.js
```
What is this file for?

waharnum replied:
Whoops - that was an accidental commit of a `find`-command file I generated to inventory the modified files for when I was updating the copyright date. Removed.

cindyli commented on src/html/chartAuthoringInterfaceTemplate.html:
```diff
@@ -1,5 +1,13 @@
 <div class="floec-chartTitle floe-chartTitle"><p class="floec-inlineEdit-text"></p></div>
 <div class="floec-chartDescription floe-chartDescription"><p class="floec-inlineEdit-text"></p></div>
+<div class="floec-sonifierControls floe-sonifierControls">
+<a href="#" class="floec-sonifierPlay floe-sonifierPlay">
```
Would be nice to have the nested indents.

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2015 OCAD University
+Copyright 2016 OCAD University
```
The years should be 2015-2016

cindyli replied:
Only the new files added in 2016 use 2016 only.

waharnum replied:
Corrected.

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -133,31 +153,77 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+                args: ["{that}", "{that}.chartAuthoringInterface.pieChart.pie.paths"]
+            }
+        },
+        model: {
```
@waharnum, seems to me the model paths of `activeRowId` and `activeSliceId` are better belong to the sub-component `pieChart` where `activeRowId` will be passed down to its `legend` sub-component and `activeSliceId` will be passed down to `pie` sub-compnent. 

The work flow is like, the model change at the `sonifier.model.currentlyPlayingData` is model-relayed to the `pieChart.model.activeRowId` and `pieChart.model.activeSliceId`` (what`updateActiveElements()`currently does).`pie`and`legend``` components respond to those model changes to highlight the corresponding pie slice and legend row. Let me know if this makes sense. Thanks.

waharnum replied:
Hi @cindyli - I think this makes sense as a delegation of responsibility and will better decouple the components involved. Given @amb26's feedback on further modelization of the play highlighting behaviour, I may roll this into whatever refactoring I do to support that. Make sense?

cindyli replied:
Yes, does make sense. Thanks.

cindyli commented on src/js/d3Utils.js:
```diff
@@ -23,6 +23,37 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+
+    // Given a selection of D3 elements and an ID, returns only the elements
+    // matching that ID
+    floe.d3.filterById = function (d3Selection, currentlyPlayingDataId) {
```
`floe.d3.filterById()` and `floe.d3.filterByNotId` are almost identical. Can we combine them into one function by passing in a flag parameter? If we do so, the name of the function might need to be changed to match the intention of the flag.

waharnum replied:
Maybe: `floe.d3.filterById(d3Selection, idToFilter, invertResult)`

`invertResult` is a boolean that results in returning either all matched or all non-matched by `idTofilter`.

cindyli replied:
Sounds good. Thanks.

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -225,18 +324,20 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
             $(this).find(dataEntryValueSelector).val(currentData.value).trigger("change");
         });
+
+        that.events.onDataAppliedToDataEntryPanel.fire();
```
Noticed this event is only used in the test. It probably can be removed by refactoring the test to use `func` directly instead of listening to this event.

amb26 commented on src/js/chartAuthoring.js:
```diff
@@ -185,22 +259,47 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+
+        var selection = d3Selector;
+
+        var activeDataId = that.model.currentlyPlayingData !== null ? that.model.currentlyPlayingData.id : null;
```
Nutty ternary - can just be 
that.model.currentlyPlayingData && that.currentlyplayingData.id
but even more directly
var activeDataId = fluid.get(model, "currentlyPlayingData.id");

amb26 commented on src/js/chartAuthoring.js:
```diff
@@ -185,22 +259,47 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+
+    floe.chartAuthoring.updateActiveElements = function (that) {
+        var currentlyPlayingData = that.model.currentlyPlayingData;
+        var currentlyPlayingDataId = currentlyPlayingData !== null ? currentlyPlayingData.id : null;
```
As above

amb26 commented on src/js/d3Utils.js:
```diff
@@ -23,6 +23,27 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+     * - "undefined" if neither of these exist
+     */
+    floe.d3.idExtractor = function (d) {
+        return d.id !== undefined ? d.id : d.data !== undefined ? d.data.id !== undefined ? d.data.id : undefined : undefined;
```
More intelligible: return fluid.get(d, "id") || fluid.get(d, "data.id")

waharnum replied:
Definitely more intelligible (and a useful reminder to me of the principle of "if you keep doing something over and over, check to make sure a function doesn't exist in Infusion for it")

amb26 commented on src/js/d3Utils.js:
```diff
@@ -23,6 +23,27 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+
+    // Given a selection of D3 elements and an ID, returns only the elements
+    // matching that ID
+    floe.d3.filterById = function (d3Selection, idToFilter, invertResult) {
```
We should be able to eliminate this function too by modelising further. Our records of which UI elements are associated with the playing element should ALSO be stored in the model or at least somewhere in the component. Then we can just use standard functional utilities to process them.

waharnum replied:
The challenge I'm having in thinking about how best to implement this is the D3 piece. The basic "get DOM elements associated with the ID of this particular data point" functionality is enabled by D3 and its approach to binding information about data to DOM nodes and maintaining object constancy through keys. So as stands, at some point in the implementation we have to "ask" D3 to tell us which elements in the legend/pie match/don't match the ID associated with a data point.

That said, I'll hack on this further and see what I can come up with that limits the role of D3's paradigms further in favour of Infusion's.

waharnum replied:
I've done an initial refactor based on @cindyli's suggestion of pushing the highlight behaviour down to the pie/legend, and I think this gets us closer to what's desirable. Will continue along these lines.

cindyli commented on src/js/unitDivisorSonificationStrategy.js:
```diff
@@ -0,0 +1,257 @@
+                funcName: "floe.chartAuthoring.sonifier.unitDivisorSonificationStrategy.dataEntriesToSonificationData",
+                args: "{that}"
+            },
+            "unitDivisor10xSonificationStrategy": {
```
Is there a need to define 3 strategies by calling the same function with different units? What do you think about defining one invoker that accepts an "unit" argument? This way I think would give more control to users. Also, `unitDivisor5xSonificationStrategy` and `unitDivisor1xSonificationStrategy` don't seem to be used at all.

waharnum replied:
I'm fine with that - those `1x` / `5x` invokers are there as examples. If it makes more sense to have one invocation with a `unit` argument that's straightforward.

amb26 replied:
Brilliant suggestion, @cindyli !

cindyli commented on src/js/unitDivisorSonificationStrategy.js:
```diff
@@ -0,0 +1,257 @@
+    // Given an object in the style of floe.chartAuthoring.dataEntryPanel.model.dataEntries,
+    // convert it to an array of objects in the style used by the sonification components,
+    // maintaining object constancy by using the dataEntry object name as the key
+    floe.chartAuthoring.sonifier.unitDivisorSonificationStrategy.dataEntriesToSonificationData = function (that) {
```
This can be done more declaratively: 

```
invokers: {
    "sonifyData": {
        changePath: "sonifiedData",
        value: "@expand:{that}.unitDivisorStrategy(10)"
    }
}
```

waharnum replied:
Awesome - I was trying to refactor that to be declarative earlier but couldn't get it to work. Will try this approach.

amb26 commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,257 @@
+        //
+        // The check for undefined prevents an error if a stop has been called
+        // but this function is queued or in the midst of execution
+        var noteDuration = (data !== undefined) ? data.notes.totalDuration : 0;
```
fluid.get(data, "notes.totalDuration") || 0

cindyli commented on tests/js/chartAuthoringTests.js:
```diff
@@ -155,8 +159,50 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+
+            var expectedData = idx < floe.tests.chartAuthoring.updateDataSet.length ? floe.tests.chartAuthoring.updateDataSet[idx] : expectedData = {label: "", value: ""};
+
+            jqUnit.assertEquals("Testing updated dataEntryPanel UI - displayed label at position " + idx + " matches updated data set label", $(this).find(dataEntryLabelSelector).val(), String(expectedData.label));
```
At line 179, 180, do you think it's possible to replace `$(this).find` with `that.locate()`? The latter ensures the selector lookup is within `that.container`, meaning it would exclude tags on the same page with the same selectors but not inside the scope of `that.container`.

waharnum replied:
While my memory is a bit hazy on the exact conversation, I think we discussed a similar issue when implementing the original `updateDataEntryPanelFromDataSet` functionality (which this is testing) back in November. The way the test looping of the `dataEntry` elements works is modelled on how the function being tested works in terms of getting all the individual `dataEntry` components from the `dataEntryPanel` and performing some action on each of them. I think [this is the final commit from November](https://github.com/waharnum/chartAuthoring/commit/6ad54c317c5c6db2c9821b8a226bf3c59f4222c3) after we looked at the issue closely and concluded that using `$(this).find)` was the best option, given Infusion doesn't have a method at the moment to easily loop over a set of `dynamicComponents`.

I think this is also already scoped to `that.container` because the original set of jQuery elements being looped by `dataEntries.each` is found by `that.chartAuthoringInterface.dataEntryPanel.locate("dataEntry");`

cindyli replied:
You are right, @waharnum. I didn't read enough context that `dataEntries` were generated from dynamic components. My bad. Thanks for refreshing my memory.

waharnum replied:
No worries @cindyli - I had to go back and refresh my own memory as to why the original functionality had been implemented with this odd mix of initial `that.locate()` then jQuery directly, but at least there was a reason for it. :smile:

amb26 replied:
We should have a chat about how we can clean this up - similar to the main strategy comment above about how to deal with links to data at the D3 level, it's a reasonable responsibility for a component to store a data structure which indirects from data onto DOM ids responsible for that data (most likely allocated using fluid.allocateSimpleId). Whilst we can't explicitly iterate over dynamicComponents, we can have them "report for duty on creation" by broadcasting onCreate listeners to them via distributeOptions, which will allow us to build up the data structure above.

cindyli replied:
@waharnum, would be nice to create a JIRA for @amb26's suggestion above. Thanks.

cindyli commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,257 @@
+        playbackOptions: {
+            // Additional gap of silence between individual data points when
+            // playing a sequence
+            gapDuration: 1,
```
Is `gapDuration` in second or other unit form? Can you add a comment for it? Would be nice if there's a test for it.

cindyli commented on tests/js/sonifierTests.js:
```diff
@@ -0,0 +1,307 @@
+            },
+            // We kick off the next test when this one finishes
+            listeners: {
+                "onSonificationStopped": {
```
Please add a namespace for this listener.

cindyli commented:
Can you add tests for play, pause buttons as well as the highlighting of the pie and the legend? Thanks.

waharnum commented:
@cindyli Tests added for the parts of `chartAuthoring.js` that intersect with the sonifier (play/stop, highlighting while playing).

waharnum commented:
I'd like to ask for some feedback on https://github.com/waharnum/chartAuthoring/compare/FLOE-437...waharnum:FLOE-437-modelexperiment?diff=split&expand=1&name=FLOE-437-modelexperiment

This is the results of some initial thinking about how we might draw brighter lines around the usage of D3's paradigms vs. Infusion's in these components. At a high level, it aims to achieve the following:
- confine D3 to DOM manipulation only when this takes place as part of D3's core enter / update / exit approach to binding data to DOM elements (use D3 for what it's good at and only that, as previously discussed)
- manipulate the DOM with Infusion / jQuery in other circumstances, such as when highlighting DOM elements in response to playback
- provide more effective bridging between Infusion's paradigms and D3's by storing key-based references (using the same `data.id` used by D3 to keep the DOM elements and their bound data in sync as the key) to all D3-bound DOM elements in the model in this style: `d3Elements.pieSlices.{id}` / `d3Elements.legendRows.{id} : DOMElement`; this gives us options for referencing DOM elements created via D3 without having to use D3 directly, along with allowing us to listen for changes to the model path, etc

I'm not sure if the above approach is sensible or simply convoluted. It also may be inappropriate to store DOM element references directly on the model like this (an alternative might be to store only the IDs, with corresponding lookup methods to find the matched DOM element).

_If the approach seems sensible at a high level I can continue refactoring towards this approach as part of this PR. If it doesn't and it needs more thinking, I'd like to propose we consider creating a separate issue to track and resolve it under the heading `More Effectively Isolate D3` or similar; I don't think the existing code's separation between Infusion and D3 is substantially muddied by the current FLOE-437, though others may disagree._

Note that there are also some issues that arise from limitations Infusion's version of jQuery has around manipulating SVG elements that I've dealt with for the moment by redundant use of the `Element.classList` API.

amb26 commented:
Thanks, @waharnum - I think the referenced pull looks like a great start. I think there is interesting room for consolidation between the functions which indirect by "data id" - for example "toggleCSSClassByDataId" and the testing function. These can become generalised factories for "doing X by data id" in a model-directed way. We should have a call next week to talk over what's going in this area. Referencing your comment above, and also mine attached to the code just now, storing just IDs attached to DOM elements (presumably allocated using fluid.allocateSimpleId()) and indirecting on them with lookup methods, is a reasonable approach.

waharnum commented:
Thanks @amb26 - I've been thinking a little about the consolidation myself and tried another approach late in day Friday that's probably closer to what you're talking about: https://github.com/waharnum/chartAuthoring/compare/FLOE-437...waharnum:FLOE-437-modelexperiment-2 - the `locateDataBoundElementsByKey` is the main difference, although it requires advance "registration" of any `selectors` in the component that will contain D3-bound elements to do its work.

I think that what you're describing probably looks something like this as a model structure:

```
model: {
  dataIdsCollection: {
    // Each dataId may have one or more DOM elements associated with it.
    "dataId1": ["fluid-simpleId1", "fluid-simpleId2"],
    "dataId2": ["fluid-simpleId3"]
  }
}
```

I'd make that part of the model of `floe.d3ViewComponent`, along with suitable factory functions for indirection by dataId. A child component of `d3ViewComponent` such as the pie or legend is then responsible for the following additional tasks when it uses D3 for DOM creation and manipulation:
- assigning a unique `id` to the DOM element with `fluid.allocateSimpleId()`
- associating that ID with the data ID at the model level

Any DOM-based operations that don't need to work directly through D3 data binding approach (such as the initial binding, or creating or updating elements based on the dataset) can then be done through the model. Does that sound sensible?

waharnum commented:
@amb26 / @cindyli  / @colinbdclark  - given timelines (next week is demo time for this work, and I need to spend some time writing about the sonification and chart authoring work for the ILDH, with time to have that reviewed) and the existing size of the PR, can I propose the following?
- I'll do further implementation and refactoring of the model approach discussed above for allowing us to manipulate D3-bound DOM elements in a model-directed way, so that we're not unnecessarily hitting the DOM directly with D3. I think I can complete the work this afternoon.
- further refinements can be the subject of new JIRA issues

Does that sound acceptable? I'm time-conscious about the need to merge this PR into master and move on to tackle the writing and summarizing.

waharnum commented:
On another note, @jobara said during the FLOE planning meeting that it would be useful to produce some "best practices" writeup from this conversation about working with Infusion + external code libraries that have functionality overlap - I think this would be valuable (if nothing else to make some of my own learning from this discussion concrete). 

Off the top of my head the big points look something like this:
1. support sharing any data created or used by code written with the library with Infusion's own model system or a similar means (as is currently done with the dataset)
2. support Infusion's "knowledge" of any special relationships created with the DOM by code using the external library
3. flowing out points 1 & 2, do not use functionality of the external library when comparable functionality can be supported by Infusion's own paradigms

amb26 commented:
@waharnum sounds great to me - as far as I'm concerned the review on this branch has been "advisory" for a while and the work described could have been spun off under separate JIRAs
The "working with external libraries" writeup sounds marvellous

waharnum commented:
I've done an implementation with test coverage that converts the existing functionality around managing DOM elements previously bound with data by D3 to use the indirect model-based approach previously discussed, so would like to request another round of review for this.

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -133,27 +181,43 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
                 args: ["{that}", []]
             }
         },
+        model: {
```
This `model` block can be removed as those model paths have been moved to the `pie` and `legend` components.

cindyli commented on src/js/d3ViewComponent.js:
```diff
@@ -114,11 +139,103 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+        var keyPath = "dataKeys." + d3Key;
+        var elementIds = fluid.get(that.model, keyPath);
+        // Create an empty array at the path if the key isn't currently defined
+        if (elementIds === undefined) {
```
Line 156 - 159 can be simplified to `elementIds = elementIds || [];`

amb26 commented on src/js/d3ViewComponent.js:
```diff
@@ -114,11 +139,101 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+    // element with a unique ID and the component, updates the key's
+    // value (an array of IDs) to include that ID
+    floe.d3ViewComponent.addElementIdToDataKey = function (d3Key, idToAdd, that) {
+        var keyPath = "dataKeys." + d3Key;
```
Even more direct would be simply

var keyPath = ["dataKeys", d3Key, idToAdd],
that.applier.change(keyPath, true)

and the corresponding "DELETE" change in the remove function

Always use hashes unless there is a fantastically good reason to use arrays (that is, that you really require ordering)

waharnum replied:
Good change. I have an unhealthy default preference for arrays over hashes that I think is a legacy of too much Java at a formative age.

amb26 replied:
We have all been weaning ourselves off the same unhealthy preference in a very long process.... as the last 7 years of framework development shows, where so many of our original configuration elements were formulated as arrays (distributeOptions, modelListeners, etc.). Part of array-ism is the "save the nanoseconds" movement, but the other part of it is "not planning for design extension" - arrays are very hard to share amongst multiple creators since their indices can't be depended on to be stable.

cindyli commented on src/js/d3ViewComponent.js:
```diff
@@ -114,11 +139,103 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+    // class on for any elements matching the ID and makes sure it's turn off
+    // for any elements not matching it
+    floe.d3ViewComponent.toggleCSSClassByDataId = function (id, toggleClass, that) {
+        var associatedElements = that.getElementsByDataKey(id);
```
This function finds the matched ones by id first, then iterate thru `that.model.dataKeys` to remove all matched ones then iterate thru the rest elements. I'm thinking it might be a bit more efficient to iterate thru `that.model.dataKeys` once and toggle the class for all elements.

waharnum replied:
Yeah, I think this style of implementation is a legacy of the old D3-based filtering approach - what you suggest is better. Will implement.

cindyli commented on src/js/pie.js:
```diff
@@ -158,15 +167,37 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
 
     floe.chartAuthoring.pieChart.pie.removeSlices = function (that) {
-        that.paths.exit().remove();
+        var removedSlices = that.paths.exit();
```
The steps taken for removing slices and texts are same. Can you refactor the code to share those steps? Thanks.

waharnum replied:
I added `floe.d3ViewComponent.exitD3Elements` and refactored the pie/legend to use it (legend rows had the same pattern for removal).

amb26 replied:
Elements which merely strut and fret their hour.

Surely that should be "exeuntD3Elements"? : P

waharnum replied:
`tomorrow && tomorrow && tomorrow`?

`recordedTime.syllables.last()`?

That's all I got :(

amb26 commented on src/js/d3ViewComponent.js:
```diff
@@ -114,11 +139,92 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+        if (fluid.isPlainObject(elementIds)) {
+            var keys = Object.keys(elementIds);
+            var elemIdCollectionWithPreface = fluid.transform(keys, function (elemId) {
+                    return "#" + elemId;
```
This is not reliable in general (although it is for our ids) because of escaping rules for characters in selectors.
Rather than fabricate a selector, it would be much more direct to simply apply

var elements = $(fluid.transform(keys, fluid.byId)); 

at the point where the elements are wanted

waharnum replied:
I've used `fluid.byId` as a starting point to refactor the `getElementsByDataKeys` function and remove the `getElementIdsAsSelector` function entirely. Much tighter, thanks.

amb26 commented on src/js/d3ViewComponent.js:
```diff
@@ -114,11 +139,92 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+            matchedElements.push(selector);
+        });
+        var elements = $(matchedElements.join(","));
+        return elements;
```
See above - these two functions can be replaced by the line quoted

cindyli commented on src/js/d3ViewComponent.js:
```diff
@@ -114,11 +139,93 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+    // for any elements not matching it
+    floe.d3ViewComponent.toggleCSSClassByDataId = function (id, toggleClass, that) {
+        // Get all D3-bound elements
+        var allElements = floe.d3ViewComponent.getElementsByDataKeys(Object.keys(that.model.dataKeys), that);
```
`fluid.keys()` can be used to replace `Object.keys()`.

cindyli commented on src/js/d3ViewComponent.js:
```diff
@@ -114,11 +139,93 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+
+        allElements.each(function (idx, elem) {
+            var dataId = elem.__data__.id;
+            if (id === dataId) {
```
line 211-219 can be simplified to:

```
var shouldHighlight = id === dataId ? true : false;
$(elem).toggleClass(cssClass, shouldHighlight);
elem.classList[shouldHighlight ? "add" : "remove"](toggleClass);
```

I renamed `toggleClass` to `cssClass` to avoid the confusion with the in-use jQuery `toggleClass()` function.

cindyli replied:
With the new implementation, `getElementsNotMatchingDataKey()` is no longer needed.

cindyli commented on src/js/chartAuthoring.js:
```diff
@@ -123,6 +153,24 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+                                model: {
+                                    "dataSet": "{dataEntryPanel}.model.dataSet",
+                                    // We relay currentlyPlayingData and isPlaying to the overall component so that sonification play events can be used to change the rest of the interface
+                                    currentlyPlayingData: "{chartAuthoring}.model.currentlyPlayingData",
```
Is the model relay for `currentlyPlayingData` and `isPlaying` necessary? `isPlaying` seems only used in `sonifier` component. `currentlyPlayingData` is used by model relays for `pie` and `legend` components that can directly refer to `{sonifier}.model. currentlyPlayingData`.

waharnum replied:
Legacy of old implementation hanging around. I looked at how this is done in other cases with relaying the dataset & other "shared" information that needs to go to both the pie & legend, and refactored  so that `pieChart` has an `activeDataId` value relayed to it by the sonifier that it then relays to `activeSliceId` / `activeRowId` of its pie/legend components.

cindyli commented on src/js/sonifier.js:
```diff
@@ -0,0 +1,260 @@
+                args: ["{that}"]
+            },
+            // Calls the default sonification strategy
+            "defaultSonificationStrategy": "{that}.unitDivisorStrategy(10)"
```
Does it make sense to define the number `10` as a component option so it's user configurable?

cindyli replied:
Just noticed this invoker is not actually being used. The similar invoker that gets used is `sonifyData` defined in `unitDivisorSonificationStrategy.js`. This invoker can be removed and the above comment applies to `sonifyData`.

cindyli commented on tests/js/chartAuthoringTests.js:
```diff
@@ -155,8 +212,93 @@ https://raw.githubusercontent.com/fluid-project/chartAuthoring/master/LICENSE.tx
+    // tool
+    floe.tests.chartAuthoringTester.verifyDatapointPlay = function (that) {
+        jqUnit.assertDeepEq("currentlyPlayingData is properly relayed between chartAuthoring component and sonifier subcomponent", that.chartAuthoringInterface.sonifier.model.currentlyPlayingData, that.model.currentlyPlayingData);
+        jqUnit.assertDeepEq("activeRowId is the id of the currently playing data in the sonifier", that.chartAuthoringInterface.sonifier.model.currentlyPlayingData.id, that.chartAuthoringInterface.pieChart.legend.model.activeRowId);
```
At line 281-284, can we also verify the actual ids, such as when the first data point is playing, the id is 0, and so on.

waharnum replied:
@cindyli can you clarify what's meant by "actual ID" in this context? Do you mean the DOM element IDs?

waharnum commented:
@cindyli this should be reviewable again
